Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm

```diff
@@ -1135,17 +1137,40 @@ static OSStatus sampleCallback(CMBufferRef buffer, void* refcon)
     assignImageBuffer(PAL::CMSampleBufferGetImageBuffer(sample));
 }
 
+void VideoMediaSampleRenderer::notifyFirstFrameAvailable(Function<void(const MediaTime&, double)>&& callback)
+{
+    assertIsMainThread();
+
+    m_hasFirstFrameAvailableCallback = WTFMove(callback);
+}
+
 void VideoMediaSampleRenderer::notifyWhenHasAvailableVideoFrame(Function<void(const MediaTime&, double)>&& callback)
 {
     assertIsMainThread();
 
     m_hasAvailableFrameCallback = WTFMove(callback);
+    m_notifyWhenHasAvailableVideoFrame = !!m_hasAvailableFrameCallback;
 }
 
 void VideoMediaSampleRenderer::notifyHasAvailableVideoFrame(const MediaTime& presentationTime, double displayTime, FlushId flushId)
 {
-    assertIsCurrent(dispatcher());
+    assertIsCurrent(dispatcher().get());
 
+    if (!m_notifiedFirstFrameAvailable) {
+        LOG(Media, "VideoMediaSampleRenderer::notifyHasAvailableVideoFrame hasFirstFrameAvailable");
+        callOnMainThread([weakThis = ThreadSafeWeakPtr { *this }, flushId, presentationTime, displayTime] {
+            if (RefPtr protectedThis = weakThis.get(); protectedThis && flushId == protectedThis->m_flushId) {
+                assertIsMainThread();
+                if (protectedThis->m_hasFirstFrameAvailableCallback)
+                    protectedThis->m_hasFirstFrameAvailableCallback(presentationTime, displayTime);
+            }
+        });
+        m_notifiedFirstFrameAvailable = true;
+        return;
```

Comment:
Should we remove the `return` here?
This would allow to make `m_hasFirstFrameAvailableCallback ` and  `m_hasAvailableFrameCallback` callbacks totally independent.

This would allow the `m_hasFirstFrameAvailableCallback` callback from MediaPlayer to not care about gathering of metadata, which might be easier to read.

Alternatively, we could return early here and call `m_hasAvailableFrameCallback` right after the call to `m_hasFirstFrameAvailableCallback`.

Also, we could remove passing presentationTime & displayTime to `m_hasFirstFrameAvailableCallback` given these parameters are not used.